File: 50-strukture/binarno-stablo-pretrage.js
class Cvor {
  constructor(vrednost) {
    this.vrednost = vrednost
    this.levo = null
    this.desno = null
  }
}

class Stablo {
  constructor() {
    this.koren = null
  }

  dodaj(vrednost) {
    const cvor = new Cvor(vrednost)

    if (!this.koren) {
      this.koren = cvor
      return
    }

    let trenutniCvor = this.koren
    while (trenutniCvor) {
      if (vrednost < trenutniCvor.vrednost) {
        if (!trenutniCvor.levo) {
          trenutniCvor.levo = cvor
          break
        } else {
          trenutniCvor = trenutniCvor.levo
        }
      } else {
        if (!trenutniCvor.desno) {
          trenutniCvor.desno = cvor
          break
        } else {
          trenutniCvor = trenutniCvor.desno
        }
      }
    }
  }
}

const stablo = new Stablo()

stablo.dodaj(8)
stablo.dodaj(3)
stablo.dodaj(10)
stablo.dodaj(1)
stablo.dodaj(6)
stablo.dodaj(14)
stablo.dodaj(4)
stablo.dodaj(7)
stablo.dodaj(13)

console.log(stablo)

File: 50-strukture/binarno-stablo-pretrage.py
class Cvor:
    def __init__(self, val):
        self.l_dete = None
        self.d_dete = None
        self.data = val


def dodaj(koren, cvor):
    if koren is None:
        koren = cvor
    else:
        if koren.data > cvor.data:
            if koren.l_dete is None:
                koren.l_dete = cvor
            else:
                dodaj(koren.l_dete, cvor)
        else:
            if koren.d_dete is None:
                koren.d_dete = cvor
            else:
                dodaj(koren.d_dete, cvor)


def stampaj_poredjano(koren):
    if not koren:
        return
    stampaj_poredjano(koren.l_dete)
    print koren.data
    stampaj_poredjano(koren.d_dete)


def stampaj_neredjano(koren):
    if not koren:
        return
    print koren.data
    stampaj_neredjano(koren.l_dete)
    stampaj_neredjano(koren.d_dete)


stablo = Cvor(3)
dodaj(stablo, Cvor(7))
dodaj(stablo, Cvor(1))
dodaj(stablo, Cvor(5))

print "poredjano:"
stampaj_poredjano(stablo)

print "neredjano:"
stampaj_neredjano(stablo)
